core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -576,13 +616,28 @@ private[spark] class BlockManagerInfo(
             s" (size: ${Utils.bytesToString(diskSize)})")
         }
       }
-      if (!blockId.isBroadcast && blockStatus.isCached) {
-        _cachedBlocks += blockId
+
+      if (!blockId.isBroadcast) {
+        if (!externalShuffleServiceEnabled || !storageLevel.useDisk) {
```

Comment:
This logic reads a bit funny although it looks correct. How about:

```
if (blockExists && externalShuffleServiceEnabled && storageLevel.useDisk) {
  _exclusiveCachedBlocks -= blockId
} else {
  _exclusiveCachedBlocks += blockId
}
```

Reply:
For me reading my version is easier: the block is an exclusive block when external shuffle service is disabled or it is enabled but not block is in the disk. Although if it is not exclusive currently but it was known before (existed) than to be in the safe side (as it could have been exclusive) we should remove it from the exclusive blocks.  

Also it uses the Set operation (add/remove) rarely. So I would like to keep this.